mapflow/functional/layer_utils.py

```diff
@@ -279,6 +280,48 @@ def add_layer(self, layer: Optional[QgsMapLayer]) -> None:
         else:  # assume user opted to not use a group, add layers as usual
             self.project.addMapLayer(layer)
 
+    def add_layer_by_order(self, layer: Optional[QgsMapLayer], order) -> None:
+        if not layer:
+            return
+        self.layer_group = self.layer_tree_root.findGroup(self.settings.value('layerGroup'))
+        if self.add_layers_to_group:
+            if not self.layer_group:  # сreate a layer group
+                self.layer_group = self.layer_tree_root.insertGroup(0, self.plugin_name)
+                # A bug fix, - gotta collapse first to be able to expand it
+                # Or else it'll ignore the setExpanded(True) calls
+                self.layer_group.setExpanded(False)
+                self.settings.setValue('layerGroup', self.plugin_name)
+                # If the group has been deleted, assume user wants to add layers to root, memorize it
+                self.layer_group.destroyed.connect(lambda: setattr(self, 'add_layers_to_group', False))
+                # Let user rename the group, memorize the new name
+                self.layer_group.nameChanged.connect(lambda _, name: self.settings.setValue('layerGroup', name))
+            # To be added to group, layer has to be added to project first
+            self.project.addMapLayer(layer, addToLegend=False)
+            # Explcitly add layer to the position 0 or else it adds it to bottom
+            self.layer_group.insertLayer(order, layer)
+            self.layer_group.setExpanded(True)
+        else:  # assume user opted to not use a group, add layers as usual
+            self.project.addMapLayer(layer)
+
+    def add_basemap_layer(self, basemap, layer):
+        # Remove layer if already exists
+        for l in self.iface.mapCanvas().layers():
+            if basemap in l.dataProvider().dataSourceUri():
+                QgsProject.instance().removeMapLayers([l.id()])
+        # Append each basemap layer to a list and add new ones at the top of a group
+        basemaps = []
+        for l in self.iface.mapCanvas().layers():
+            if "{z}" or "http" in l.dataProvider().dataSourceUri():
```

Comment:
what if data source is a file "httpdata.goeojson"?
Also, this puts the preview layers on top of the processing result layers, which is not good.

I suggest not to check all the layers' types (the user's project can contain different layers, we should not manage them).
Let's do it this way:
- add "preview_layers=[]" in mapflow class __init__
- when adding a preview layer, first check if the layers in the list are still present in the project, and remove the removed layers from the list, than add the new layer to the top of the list and to the top of the topmost layer from the list.

So we will not try to manage the user's layers, and will not put the preview layers on top of the result layers (it is bad because preview covers everything, and we want to see results on top of preview).
If user places one of his layers over our previews, I think he knows what he is doing, we don't need to interfere.